### ISSUE TITLE
Dockerfile.gpu improvement

### ISSUE THREAD

**Issue by rsdmse** [microsoft/LightGBM#3286]
## Summary

A Dockerfile for the CLI is prepared and tested on GPU. Improvements are made to reduce the final image size compared to the current Dockerfile.gpu.

## Motivation

The final image size can be drastically reduced from 2.6 GB (Dockerfile.gpu but without python) to 111 MB. (Here I just compare CLI-only images. Adding conda and the necessary python packages would increase the final image size in both cases by roughly 1 GB; i.e. 3.6 GB vs 1.1 GB.)

## Description

Problems with the current Dockerfile.gpu:

1. The cleanup commands in Dockerfile.gpu do not reduce the image size. They must be in the same RUN statement as the apt-get/conda installation commands.

1. The use of `cuda*-cudnn*-devel` as the production base image is unnecessary.

I have prepared a CLI-only Dockerfile [here](https://github.com/uvarc/rivanna-docker/blob/master/lightgbm/2.3.1/Dockerfile). LightGBM version 2.3.1 is built from source with CUDA 10.2 and cuDNN 8 in a multistage build. The CLI binary and necessary libraries are copied to the final stage.

Note that the `cuda10.2-cudnn8-devel` image is almost 2.5 GB by itself. If this were used as the production image, I estimate the image size to be around 2.6 GB. Using `cuda10.2-base` as the production image, I obtained an image size of only 111 MB.

The container was benchmarked on our HPC platform (Rivanna cluster, University of Virginia) using [this tutorial](https://lightgbm.readthedocs.io/en/latest/GPU-Tutorial.html#dataset-preparation). It took 33 seconds for 50 iterations on a GPU, compared to 92 seconds without GPU.

It is not clear to me whether CUDA math libraries/NCCL/cuDNN are needed. They were not needed for the above tutorial at runtime. But if yes, we could use `runtime` as the production image. Its estimated image size of 1.2 GB would still be less than half of the current image size of 2.6 GB.

## References

Please feel free to test our container: https://hub.docker.com/r/uvarc/lightgbm

The tag must be provided explicitly since we do not use the `latest` tag:
```
docker pull uvarc/lightgbm:2.3.1
```

**Comment by StrikerRUS** [comment]
Hi @rsdmse !

Thanks a lot for your investigation! Please feel free to create a PR with the described improvements.

Speaking about `runtime` version of CUDA images, please check the following thread: https://github.com/microsoft/LightGBM/issues/3040#issuecomment-669288968.

**Comment by rsdmse** [comment]
Thank you! I'll create a `Dockerfile.cli-gpu` (CLI-only) to distinguish it between `Dockerfile.gpu` (CLI + Python).

**Comment by mirekphd** [comment]
> It is not clear to me whether CUDA math libraries/NCCL/cuDNN are needed. 

CuDNN - not and will not. The rest - not, but will be once the PR adding CUDA support is merged in.

**Comment by rsdmse** [comment]
Thanks for the clarification. In that case I think it would make more sense to use `nvidia/opencl` rather than `nividia/cuda`, at least for now. I just built an image with the former and tested it on our machine. As expected, it has exactly the same performance. The image size is reduced by another 7 MB to 104 MB.

**Comment by mirekphd** [comment]
One peculiarity I noted is the fact that the popular packages which are either device-agnostic or have a pre-compiled GPU variant (would that be the purpose of this Dockerfile too?) such as e.g. xgboost and tensorflow-gpu, are still compiled against CUDA 10.1, which NVIDIA has effectively abandoned... 8 months ago (see [nvidia/cuda](https://hub.docker.com/r/nvidia/cuda/tags?page=1&name=10.1)), resulting in a literally hundred (!) unnecessary vulnerabilities for the unsuspecting users of their public images...

**Comment by rsdmse** [comment]
We tend to containerize GPU applications for users on our HPC cluster. A user requested LighGBM recently and I came across the `dockerfile.gpu` in this repo. After some testing I think the 0.1 GB version works just as well as the 2.6 GB version.

We also have a customized dockerfile for tensorflow, but that's mainly because we wanted to use python 3.7 instead of their default 3.6. It's largely based on the official dockerfile and the final image size is very similar. The base image is `cuda:base` which is the leanest among all 3 (`base`, `runtime`, `devel`).

The base image and proper cleanup can make a huge difference to the final image size.

**Comment by rsdmse** [comment]
I just tested that `distroless` works too. Again, same performance on GPU. The image size is further reduced to merely **14 MB**. Please feel free to test this out:
```
docker pull uvarc/lightgbm:2.3.1-distroless
```
